File: android/src/main/kotlin/co/quis/flutter_contacts/crud/models/contact/ContactMetadata.kt
package co.quis.flutter_contacts.crud.models.contact

data class ContactMetadata(
    val properties: Set<String>,
    val accounts: List<Map<String, String>> = emptyList(),
) {
    companion object {
        fun fromJson(json: Map<String, Any?>): ContactMetadata {
            val propertiesList = json["properties"] as? List<*> ?: emptyList<Any>()
            val properties = propertiesList.map { it as String }.toSet()
            val accountsList = json["accounts"] as? List<*> ?: emptyList<Any>()
            val accounts =
                accountsList.mapNotNull { account ->
                    (account as? Map<*, *>)?.let { map ->
                        val name = map["name"] as? String
                        val type = map["type"] as? String
                        if (name != null && type != null) {
                            mapOf("name" to name, "type" to type)
                        } else {
                            null
                        }
                    }
                }
            return ContactMetadata(properties = properties, accounts = accounts)
        }

        fun fromPropertiesSet(
            properties: Set<String>,
            accounts: List<Map<String, String>> = emptyList(),
        ): ContactMetadata = ContactMetadata(properties = properties, accounts = accounts)
    }

    fun toJson(): Map<String, Any?> {
        return mapOf(
            "properties" to properties.sorted().toList(),
            "accounts" to accounts,
        )
    }
}

File: ios/Classes/crud/models/contact/ContactMetadata.swift
import Foundation

struct ContactMetadata: Equatable, Hashable {
    let properties: Set<ContactProperty>
    let accounts: [Account]

    func toJson() -> Json {
        [
            "properties": [String](properties.map { $0.rawValue }.sorted()),
            "accounts": accounts.map { $0.toJson() },
        ]
    }

    static func fromJson(_ json: Json) -> ContactMetadata {
        let properties = Set((json["properties"] as? [String] ?? []).compactMap(ContactProperty.init))
        let accounts = (json["accounts"] as? JsonArray ?? []).map(Account.fromJson)
        return ContactMetadata(properties: properties, accounts: accounts)
    }
}
